kernel/shared/src/main/scala/cats/effect/kernel/Resource.scala

```
@@ -757,87 +760,327 @@ object Resource extends ResourceInstances with ResourcePlatform {
+    new ResourceMonadCancel[F] {
+      def F = F0
+      def rootCancelScope = F0.rootCancelScope
+    }
+}
+
+private[effect] trait ResourceHOInstances4 extends ResourceHOInstances5 {
+  implicit def catsEffectMonadErrorForResource[F[_], E](
+      implicit F0: MonadError[F, E]): MonadError[Resource[F, *], E] =
+    new ResourceMonadError[F, E] {
+      def F = F0
+    }
+}
+
+private[effect] trait ResourceHOInstances5 {
   implicit def catsEffectMonadForResource[F[_]](implicit F0: Monad[F]): Monad[Resource[F, *]] =
     new ResourceMonad[F] {
       def F = F0
     }
+}
 
+abstract private[effect] class ResourceFOInstances0 extends ResourceFOInstances1 {
+  implicit def catsEffectMonoidForResource[F[_], A](
+      implicit F0: Monad[F],
+      A0: Monoid[A]): Monoid[Resource[F, A]] =
+    new ResourceMonoid[F, A] {
+      def A = A0
+      def F = F0
+    }
+}
+
+abstract private[effect] class ResourceFOInstances1 {
   implicit def catsEffectSemigroupForResource[F[_], A](
       implicit F0: Monad[F],
       A0: Semigroup[A]): ResourceSemigroup[F, A] =
     new ResourceSemigroup[F, A] {
       def A = A0
       def F = F0
     }
+}
 
-  implicit def catsEffectSemigroupKForResource[F[_], A](
-      implicit F0: MonadCancel[F, Throwable],
-      K0: SemigroupK[F],
-      G0: Ref.Make[F]): ResourceSemigroupK[F] =
-    new ResourceSemigroupK[F] {
-      def F = F0
-      def K = K0
-      def G = G0
+// TODO the rest of the instances
+abstract private[effect] class ResourceMonadCancel[F[_]]
+    extends ResourceMonadError[F, Throwable]
+    with MonadCancel[Resource[F, *], Throwable] {
+  implicit protected def F: MonadCancel[F, Throwable]
+
+  def canceled: Resource[F, Unit] =
+    Resource.eval(F.canceled)
+
+  def forceR[A, B](fa: Resource[F, A])(fb: Resource[F, B]): Resource[F, B] =
+    Resource.applyFull(
+      _(fa.use_ !> fb.allocated).map(_.map(fin => (_: Resource.ExitCase) => fin)))
```

I think `fa.attempt.productR(fb)` would match the described semantics of `forceR` better. Currently the scope of `fa` is closed, even on success, prior to `fb`.

---

Whether or not to close the scope of the left side in `forceR` I think is a decision we can make here. I'm not sure what's right or wrong. If we want to keep the scope open, then something like `fold` putting hte finalizers in a `Ref` and then still delegating to `forceR` on the `F` for the composition might be the right approach.

We definitely can't just use `.attempt.productR` since `F` might be `OptionT` or something like that.

---

Fair enough regarding `attempt`. Thinking about it in the context of the success case, being consistent there with `fa *> fb` seemed least surprising.

---

Spent some more time thinking about this. I think the only consistent semantic is to close the left scope. The problem here is that the left side *might* error, and if it does, its finalizers will naturally be run. However, that error will be discarded (by the semantics of `forceR`) and the right will be run. In that case, the finalizers are forcibly run early. If the left side *succeeds* and we attempt to keep the scope open, then the result will be the same from the user's perspective (the value will be discarded and the right side sequenced), but now the scope is mysteriously open. So that's more than a bit weird.

I think it's a little simpler if we just say that the scope on the left always closes. This gives `forceR` the same semantics as `supersedeBy` from `Region` in some of the earlier CE3 prototypes, which I think is useful in its own right.

---

Hm, maybe my understanding of `forceR` is incomplete/misguided. How should one view `fa forceR fb` in relation to `fa *> fb` when `fa` and `fb` succeed (asking in a parametric context here, not specifically for `Resource`)? My current understanding, from the description of `forceR`, is that `fa forceR fb` and `fa *> fb` are equivalent when both arguments succeed.

Similarly, when I know I'm dealing with `Resource`, I'd expect `fa`'s scope to extend over `fb` in `fa *> fb` so I thought it seemed reasonable that `forceR` should behave accordingly.

In the case where `fa` errors, having it's finalizers run is also expected and consistent with `*>`, the only difference with `forceR` being that `fb` would still evaluate, but that's expected and the point of calling `forceR` in the first place.

---

> My current understanding, from the description of forceR, is that fa forceR fb and fa *> fb are equivalent when both arguments succeed.

This is an area that I think we can do better on actually *defining*. It's difficult to do generically though. Another area where this shows up is `StateT`. If I remember correctly, `StateT`'s `MonadCancel` instance discards state in one of the branches of `forceR` (I think the right?).

In a sense, `fa !> fb` is meant to mirror `(fa *> F.canceled).onCancel(fb)`, whatever that ends up being (also naively ignoring errors in `fa` in this case). This is obviously a strange situation, since the finalizer kind of lives in its own universe under *normal* circumstances (when canceled externally). Trying to figure out what this means for `Resource` is a little mind-bendy…